cordapp-example/contracts-kotlin/src/main/kotlin/com/example/contract/IOUContract.kt

```diff
@@ -30,17 +30,32 @@ class IOUContract : Contract {
      * considered valid.
      */
     override fun verify(tx: LedgerTransaction) {
-        val command = tx.commands.requireSingleCommand<Commands.Create>()
-        requireThat {
-            // Generic constraints around the IOU transaction.
-            "No inputs should be consumed when issuing an IOU." using (tx.inputs.isEmpty())
-            "Only one output state should be created." using (tx.outputs.size == 1)
-            val out = tx.outputsOfType<IOUState>().single()
-            "The lender and the borrower cannot be the same entity." using (out.lender != out.borrower)
-            "All of the participants must be signers." using (command.signers.containsAll(out.participants.map { it.owningKey }))
-
-            // IOU-specific constraints.
-            "The IOU's value must be non-negative." using (out.value > 0)
+        val command = tx.commands.requireSingleCommand<Commands>()
+
+        when (command.value) {
+            is Commands.Create -> requireThat {
+                // Generic constraints around the IOU transaction.
+                "No inputs should be consumed when issuing an IOU." using (tx.inputs.isEmpty())
+                "Only one output state should be created." using (tx.outputs.size == 1)
+                val out = tx.outputsOfType<IOUState>().single()
+                "The lender and the borrower cannot be the same entity." using (out.lender != out.borrower)
+                "All of the participants must be signers." using (command.signers.containsAll(out.participants.map { it.owningKey }))
+
+                // IOU-specific constraints.
+                "The IOU's value must be non-negative." using (out.value > 0)
+            }
+            is Commands.Destroy -> requireThat {
+                "There should be one IOU input." using (tx.inputs.size == 1)
```

Comment:
You reused `tx.inputs`, as found in the example. However this means that:

1. you are checking that there is exactly 1 input only
2. you are not sure, on this line, that this single input is an IOUState, which you seem to believe on line 51.
3. by doing that, you are preventing this `Destroy` operation from being atomically combined with something else, for instance a payment. Wouldn't that be great to be able to achieve that: a payment that clears the IOU? 
4. same remark in fact for no outputs. The payment would just change the owner of "dollar" tokens, so it would mean input(s) and output(s).
5. same remark for `.Create`. After all, the IOU could be created at the same time that dollar tokens are transferred...

Reply:
I'm a little confused about payment part of your comment. I understand that this task (in module 3) is for payback only - not payment. Cash flow task (payments) is in module 4. Or I misunderstood something?

About line 51 - assuming I want only one input and `IOUState` type, line 48, where I'm checking size, is not needed at all?

Reply:
By payment I mean that if a cash would be used as the payment to destroy an IOU, you will not be able to atomically mix the 2. 

You need to check that `tx.inputsOfType<IOUState>().size == 1`. That's what the contract cares about. It should not care about other types of states.

Reply:
but `tx.inputsOfType<IOUState>().single()` this will throw error if empty or not single, so checking `size == 1` is not needed.

Reply:
If it were only a question of failing when it's wrong, ok.
But I am also saying that you want your error messages to be clear. "There was not a single IOUState" is better than "Wrong number of elements in collection".

Reply:
good point :)